waitable_atomic/waitable_atomic.h

```diff
@@ -0,0 +1,324 @@
+          }
+        }
+        void MarkAsUnmodified() { mark_as_unmodified_ = true; }
+
+       private:
+        POINTER* parent_;
+        bool mark_as_unmodified_;
+      };
+      typedef typename std::conditional<std::is_const<POINTER>::value,
+                                        ImmutableAccessorDoesNotNotify,
+                                        MutableAccessorDoesNotify>::type type;
+    };
+
+    template <class PARENT>
+    class TemplatedScopedAccessor : private ScopedUniqueLock, public NotifyIfMutable<PARENT>::type {
+     public:
+      typedef PARENT T_PARENT;
+      typedef typename NotifyIfMutable<PARENT>::type T_OPTIONAL_NOTIFIER;
+
+      typedef typename std::conditional<std::is_const<T_PARENT>::value,
+                                        const typename T_PARENT::T_DATA,
+                                        typename T_PARENT::T_DATA>::type T_DATA;
+
+      explicit inline TemplatedScopedAccessor(T_PARENT* parent)
+          : ScopedUniqueLock(parent->data_mutex_), T_OPTIONAL_NOTIFIER(parent), pdata_(&parent->data_) {}
+
+      inline TemplatedScopedAccessor(TemplatedScopedAccessor&& rhs)
+          : ScopedUniqueLock(std::move(rhs)), T_OPTIONAL_NOTIFIER(rhs), pdata_(rhs.pdata_) {}
+
+      inline ~TemplatedScopedAccessor() {}
+
+      TemplatedScopedAccessor() = delete;
+      TemplatedScopedAccessor(const TemplatedScopedAccessor&) = delete;
+      void operator=(const TemplatedScopedAccessor&) = delete;
+
+      inline T_DATA* operator->() { return pdata_; }
+      inline T_DATA& operator*() { return *pdata_; }
+
+     private:
+      T_DATA* pdata_;
+    };
+
+    typedef TemplatedScopedAccessor<BasicImpl> MutableAccessor;
+    typedef TemplatedScopedAccessor<const BasicImpl> ImmutableAccessor;
+
+    friend class TemplatedScopedAccessor<BasicImpl>;
+    friend class TemplatedScopedAccessor<const BasicImpl>;
+
+    inline ImmutableAccessor ImmutableScopedAccessor() const { return ImmutableAccessor(this); }
+
+    inline MutableAccessor MutableScopedAccessor() { return MutableAccessor(this); }
+
+    inline void Notify() { data_condition_variable_.notify_all(); }
+
+    inline void UseAsLock(std::function<void()> f) const {
+      std::unique_lock<std::mutex> lock(T_DATA::data_mutex_);
+      f();
+    }
+
+    inline bool Wait(std::function<bool(const T_DATA&)> predicate) const {
+      std::unique_lock<std::mutex> lock(data_mutex_);
+      if (!predicate(data_)) {
+        const T_DATA& data = std::ref(data_);
+        data_condition_variable_.wait(lock, [&predicate, &data] { return predicate(data); });
+      }
+      return true;
+    }
+
+    inline void ImmutableUse(std::function<void(const T_DATA&)> f) const {
+      auto scope = ImmutableScopedAccessor();
+      f(*scope);
+    }
+
+    inline void MutableUse(std::function<void(T_DATA&)> f) {
+      auto scope = MutableScopedAccessor();
+      f(*scope);
+    }
+
+    inline void PotentiallyMutableUse(std::function<bool(T_DATA&)> f) {
```

Comment:
huh, that's an easy way out and easy way to shoot yourself. never trust your fellow programmer - something along the following lines would be safer:
```c++
void PotentiallyMutableUse(std::function<void(const T_DATA&, std::function<MutableAccessor()>)>){};
blah.PotentiallyMutableUse([](data, upgradeFn) {
   if (data.isStale()){
     auto& mutableData = *(upgradeFn());
     mutableData.moveYourAss();
   }
})
```